### PR TITLE
Add XenVM dom0 virtual shield and prepare/cleanup r8a7795 configs/code for working with the new shield  

### PR DESCRIPTION
**Pull request by xakep-amatop**

### CODE REVIEW

File: boards/shields/xen_dom0_rcar/doc/index.rst
Comment (xakep-amatop) [inline boards/shields/xen_dom0_rcar/doc/index.rst:1]: (note) I'll add documentation in a separate commit

**Comment by xakep-amatop** [comment]
there are some errors in GitHub tests
pls, postpone review of this pull until I fix them

**Comment by xakep-amatop** [comment]
> there are some errors in GitHub tests pls, postpone review of this pull until I fix them

ready for review

File: drivers/xen/Kconfig
Comment (lorc) [inline drivers/xen/Kconfig]: I'd prefer to have it `n` by default. This is to prevent issues with existing setup, which my use `xenvm` as DomU

File: soc/arm64/renesas_rcar/gen3/mmu_regions.c
Comment (lorc) [inline soc/arm64/renesas_rcar/gen3/mmu_regions.c:24]: Shouldn't you fold this change into "drivers: uart_rcar: do runtime MMIO mapping"?

I am asking, because your previous patch can break Zephyr is applied without this change here.

File: soc/arm64/renesas_rcar/gen3/mmu_regions.c
Comment (lorc) [inline soc/arm64/renesas_rcar/gen3/mmu_regions.c:36]: I believe that this change should belong to previous commit.

File: arch/arm64/core/CMakeLists.txt
Comment (lorc) [inline arch/arm64/core/CMakeLists.txt]: Ugh. Why do you hardcoding Xen version there?  

For this exact reason, there are two separate compatible entries: `compatible = "xen,xen-4.15", "xen,xen";` please use the second one if there is no hard dependency on this particular Xen version. 

Also, please take into account that 4.15 is not an actual Xen version running on a platform.

File: boards/shields/xenvm_dom0/Kconfig.defconfig
Comment (lorc) [inline boards/shields/xenvm_dom0/Kconfig.defconfig]: Is there reason why you can't use value from original SoC/Board?

File: boards/shields/xenvm_dom0/Kconfig.defconfig
Comment (lorc) [inline boards/shields/xenvm_dom0/Kconfig.defconfig]: Do you really need to enable this for all builds?

File: boards/shields/xenvm_dom0/Kconfig.defconfig
Comment (lorc) [inline boards/shields/xenvm_dom0/Kconfig.defconfig]: Is there reason why you can't use value from original SoC/Board?

File: boards/shields/xenvm_dom0/Kconfig.defconfig
Comment (lorc) [inline boards/shields/xenvm_dom0/Kconfig.defconfig]: Is there reason why you can't use value from original SoC/Board?

File: boards/shields/xenvm_dom0/Kconfig.defconfig
Comment (lorc) [inline boards/shields/xenvm_dom0/Kconfig.defconfig]: Is there reason why you can't use value from original SoC/Board?

File: boards/shields/xenvm_dom0/Kconfig.defconfig
Comment (lorc) [inline boards/shields/xenvm_dom0/Kconfig.defconfig]: I believe that you have valid reason to define all the stack sizes there

File: boards/shields/xenvm_dom0/doc/index.rst
Comment (lorc) [inline boards/shields/xenvm_dom0/doc/index.rst]: Strange wording. You can use "board config" as "initial domain". "  How about "This virtual shield allows user to build Zephyr as a Xen initial domain (Dom0). This feature is implemented as shield to allow support for any compatible platform".

File: boards/shields/xenvm_dom0/Kconfig.shield
Comment (lorc) [inline boards/shields/xenvm_dom0/Kconfig.shield:5]: Could you please add `depends on ARMV8_A`? I don't believe that this would work on CorteR or MIPS :)

File: boards/shields/xenvm_dom0/xenvm_dom0.overlay
Comment (lorc) [inline boards/shields/xenvm_dom0/xenvm_dom0.overlay]: RAM base and RAM size should be platform-dependent.

**Review by lorc — COMMENTED** [review]
This looks quite good. Thank you. But I want to ask you for two things:

- Be sure that commits are atomic. Among other things this means that I can do `git checkout` at any commit in the PR and nothing will break.
- Provide meaningful commit descriptions.  `boards: arm64: h3ulcb: add ram node to chosen` is the largest offender. I saw the changes, I saw the commit description. Looks like you are doing right thing, but I have no idea why you are doing this and how to check if this is really correct way. Please provide elaborated commit descriptions for every commit.

File: arch/arm64/core/CMakeLists.txt
Comment (xakep-amatop) [reply arch/arm64/core/CMakeLists.txt]: because CONFIG_DT_HAS_XEN_XEN_ENABLED wasn't defined during build and GitHub automation notified about that
if you want I can add extra check here too

File: boards/shields/xenvm_dom0/Kconfig.defconfig
Comment (xakep-amatop) [reply boards/shields/xenvm_dom0/Kconfig.defconfig]: different number of IRQ for h3ulcb and XENVM boards, e.g. 240 vs 500

File: boards/shields/xenvm_dom0/Kconfig.defconfig
Comment (xakep-amatop) [reply boards/shields/xenvm_dom0/Kconfig.defconfig]: sorry, I didn't investigate this and below questions, I have just copy-pasted a part of XENVM board config

File: drivers/xen/Kconfig
Comment (xakep-amatop) [reply drivers/xen/Kconfig]: good point! I'll change it to `n`

File: soc/arm64/renesas_rcar/gen3/mmu_regions.c
Comment (xakep-amatop) [reply soc/arm64/renesas_rcar/gen3/mmu_regions.c:24]: ok, I'll squash commits together

File: soc/arm64/xenvm/Kconfig.soc
Comment (firscity) [inline soc/arm64/xenvm/Kconfig.soc:12]: This commit will broke build, please add only atomic changes. All platforms (xenvm, xenvm-qemu should be buildable)

File: boards/arm64/rcar_h3ulcb_ca57/rcar_h3ulcb_ca57_defconfig
Comment (firscity) [inline boards/arm64/rcar_h3ulcb_ca57/rcar_h3ulcb_ca57_defconfig:28]: Don't think that these changes should belong to commit with such description.

File: arch/arm64/core/CMakeLists.txt
Comment (firscity) [reply arch/arm64/core/CMakeLists.txt]: Probably Zephyr DT parser does not add all compatibles to these defines. I think it will be okay to remove 4.15 from DT compatibles at all. All Zephyr drivers uses `xen,xen`. I don't think that we should update exact version every time.

File: boards/shields/xenvm_dom0/Kconfig.defconfig
Comment (firscity) [reply boards/shields/xenvm_dom0/Kconfig.defconfig]: I think that we need to add stack changes with `TODO:` as they were added to xenvm_defconfig. No one investigated exact value, that we need, it was just a WA for Gen3 boards.

**Comment by xakep-amatop** [comment]
fix everything except this comment https://github.com/xen-troops/zephyr/pull/13/files#r1137772562
working on solution...

**Comment by dsemenets** [comment]
Looks like you forget https://github.com/xen-troops/zephyr/pull/13/files#r1137782692

**Comment by xakep-amatop** [comment]
> Looks like you forget https://github.com/xen-troops/zephyr/pull/13/files#r1137782692

@dsemenets you are right, thanks
looks like, I have deleted it accidentally during cleanup

File: boards/shields/xenvm_dom0/Kconfig.shield
Comment (xakep-amatop) [reply boards/shields/xenvm_dom0/Kconfig.shield:5]: done

File: soc/arm64/renesas_rcar/gen3/mmu_regions.c
Comment (lorc) [reply soc/arm64/renesas_rcar/gen3/mmu_regions.c:36]: As I said, this change should be squashed to the previous commit

**Review by lorc — COMMENTED** [review]
There are lots of semi-related commits in this PR, which makes it hard to review.

Some of them need to be squashed together. Other - should be split up.  `boards: shields: add XEN Dom-0 shield` for example adds new DTS binding. Why are you doing this in this particular commit?

I propose to split this PR into couple of separate ones. For example, changes to PFC, UART and CPG drivers are not Xen-related and almost already merged. Just move them to a separate PR. Xen shield can go into another.

Funny thing is that PR is named   `xt rcar mmc driver integration xenvm` but there are no MMC-related changes at all :)

**Comment by xakep-amatop** [comment]
@lorc 

> Some of them need to be squashed together.

agree, there is one commit remaining that can be squashed

> Other - should be split up. `boards: shields: add XEN Dom-0 shield` for example adds new DTS binding. Why are you doing this in this particular commit?

because all stuff inside the last commit related to adding of a new shield, I see no reason for splitting it up 

> I propose to split this PR into couple of separate ones. For example, changes to PFC, UART and CPG drivers are not Xen-related and almost already merged. Just move them to a separate PR. Xen shield can go into another.

I think, it is better to leave the pull as is, because it will require more unnecessary time wasting from me and you, you need time for carefully review of everything (again) and I need time for creating another pull, moreover changes distinguish from one another by commits and  all commits will have the same appearance as after merge from different pulls as after merge only the one pull

> Funny thing is that PR is named xt rcar mmc driver integration xenvm but there are no MMC-related changes at all :)

it is historical name, content of pull-request was changed a lot during development process

File: soc/arm64/renesas_rcar/gen3/mmu_regions.c
Comment (xakep-amatop) [reply soc/arm64/renesas_rcar/gen3/mmu_regions.c:36]: done

**Comment by lorc** [comment]
> because all stuff inside the last commit related to adding of a new shield, I see no reason for splitting it up

You can go further and squash all commits into one :) 

Actually no. You need to move unrelated changes to a separate commits. For example, DTS schema may be picked up by Zephyr maintainer separately from Xen shield.

> I think, it is better to leave the pull as is, because it will require more unnecessary time wasting from me and you, you need time for carefully review of everything (again) and I need time for creating another pull, moreover changes distinguish from one another by commits and all commits will have the same appearance as after merge from different pulls as after merge only the one pull

Well, I am not forcing you. This was just my idea to speed up the review process.

> it is historical name, content of pull-request was changed a lot during development process

FYI, you can rename the PR at any time.

File: boards/arm64/rcar_h3ulcb_ca57/rcar_h3ulcb_ca57_defconfig
Comment (xakep-amatop) [reply boards/arm64/rcar_h3ulcb_ca57/rcar_h3ulcb_ca57_defconfig:28]: I have extended commit message, thanks

File: soc/arm64/xenvm/Kconfig.soc
Comment (xakep-amatop) [reply soc/arm64/xenvm/Kconfig.soc:12]: done

File: boards/arm64/rcar_h3ulcb_ca57/rcar_h3ulcb_ca57_defconfig
Comment (lorc) [reply boards/arm64/rcar_h3ulcb_ca57/rcar_h3ulcb_ca57_defconfig:28]: Commit name still says "dts: arm64: changes for r8a7795.dtsi".

Also, please take note that this is too broad for the commit name. Any patch can be described as "changes to some_file.ext"

Comment (xakep-amatop) [reply boards/arm64/rcar_h3ulcb_ca57/rcar_h3ulcb_ca57_defconfig:28]: as I have written I changed commit message and @firscity has said that changes isn't related to commit description
nobody has said nothing about commit name

**Comment by xakep-amatop** [comment]
> Actually no. You need to move unrelated changes to a separate commits. For example, DTS schema may be picked up by Zephyr maintainer separately from Xen shield.

any part of commit may be picked by anyone...

> Well, I am not forcing you. This was just my idea to speed up the review process.
 
as far as I know you and Dmytro have already reviewed all my changes, how it can speed up review process... sorry, but I have strong feeling that you are laughing at me

File: boards/arm64/rcar_h3ulcb_ca57/rcar_h3ulcb_ca57_defconfig
Comment (lorc) [reply boards/arm64/rcar_h3ulcb_ca57/rcar_h3ulcb_ca57_defconfig:28]: Commit name is a part of the commit description as a whole. Just by looking at your commit (for example when doing `git log --oneline`) tells me that you are touching only r8a7795.dtsi. Which is not true.

**Review by lorc — COMMENTED** [review]
> any part of commit may be picked by anyone...

This is technically true. But it is customary to cherry-pick commits as a whole. This helps preserving code authorship. Remember, that commits should be atomic. One commit should do one specific thing. This eases up and speeds up the review process and leads to a much cleaner and well understood history.  Also, you easily can revert some of the changes, if they cause issues. 

> as far as I know you and Dmytro have already reviewed all my changes, how it can speed up review process... sorry, but I have strong feeling that you are laughing at me

I am sorry that you have this feeling. I didn't meant harm. But, from my point of view, there are lot to do in some of the commits. As you can see, there are still ongoing conversations regarding on some of the changes. 

Also, I want to ask you to review again the following two sections:

https://www.kernel.org/doc/html/latest/process/submitting-patches.html#split-changes

https://www.kernel.org/doc/html/latest/process/submitting-patches.html#describe-your-changes

Those are the general guidelines that we (and all the open-source community) trying to follow.

Pay attention to the next statements: 

> "The point to remember is that each patch should make an easily understood change that can be verified by reviewers. Each patch should be justifiable on its own merits." 

> "Solve only one problem per patch. If your description starts to get long, that’s a sign that you probably need to split up your patch." 

> "Describe your changes in imperative mood, e.g. “make xyzzy do frotz” instead of “[This patch] makes xyzzy do frotz” or “[I] changed xyzzy to do frotz”, as if you are giving orders to the codebase to change its behaviour."

File: boards/arm64/rcar_h3ulcb_ca57/rcar_h3ulcb_ca57.yaml
Comment (lorc) [inline boards/arm64/rcar_h3ulcb_ca57/rcar_h3ulcb_ca57.yaml:8]: commit description mentions only 16MB... I do understand why you are making this change:  our board configuration was really inconsistent, but it is worth mentioning in the commit message.

**Review by lorc — APPROVED** [review]
Thank you. It is much easier to review and it is better structured overall.

I have left last comment. Please address it and then add my tag to all the commits:

`Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`

File: boards/arm64/rcar_h3ulcb_ca57/rcar_h3ulcb_ca57.yaml
Comment (xakep-amatop) [reply boards/arm64/rcar_h3ulcb_ca57/rcar_h3ulcb_ca57.yaml:8]: done

**Review by firscity — APPROVED** [review]
Good job, thank you, please add
`Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>`

**Pull request merged**
